src/collider/Geom.cpp

```diff
@@ -73,160 +68,161 @@ void Geom::CollideSphere(Sphere &sphere, void (*callback)(CollisionContact *)) c
+		stackobj curr = stack[stackpos--];
+		const SingleBVHTreeBase::Node *edgeNode = edgeBvh->GetNode(curr.edgeNode);
+		const SingleBVHTreeBase::Node *triNode = triBvh->GetNode(curr.triNode);
+
+		// NOTE:
+		// This behavior is preserved almost verbatim from existing collision
+		// code. It's very possible there are subtle bugs and inaccuracies in
+		// the BVH traversal logic that result in missed collision contacts.
+		// However, this is the most performant version of this code by a large
+		// margin, with seemingly no impact on collision robustness.
+
+		if (!edgeNode->kids[0]) {
+			// Reached edge leaf node, perform edge-triangle ray intersection
+			CollideEdgeTris(contacts, transTo, b, edgeNode->leafIndex, curr.triNode, isect_buf);
+
+			if (contacts.size() >= maxContacts) break;
+		} else if (!triNode->kids[0]) {
+			// Reached triangle leaf node, recursively descend through edge nodes
+			stack[++stackpos] = stackobj { edgeNode->kids[1], curr.triNode };
+			stack[++stackpos] = stackobj { edgeNode->kids[0], curr.triNode };
 		} else {
-			BVHNode *left = triNode->kids[0];
-			BVHNode *right = triNode->kids[1];
-			bool edgeNodeIsectsLeftChild = rotatedAabbIsectsNormalOne(edgeNode->aabb, transTo, left->aabb);
-			bool edgeNodeIsectsRightChild = rotatedAabbIsectsNormalOne(edgeNode->aabb, transTo, right->aabb);
-			//edgeNodeIsectsRightChild = edgeNodeIsectsLeftChild = true;
-			if (edgeNodeIsectsRightChild) {
-				if (edgeNodeIsectsLeftChild) {
-					// isects both. split edgeNode and try again
-					++stackpos;
-					stack[stackpos].edgeNode = edgeNode->kids[0];
-					stack[stackpos].triNode = triNode;
-					++stackpos;
-					stack[stackpos].edgeNode = edgeNode->kids[1];
-					stack[stackpos].triNode = triNode;
-				} else {
-					// hits only right child. go down into that
-					// side with same edge node
-					++stackpos;
-					stack[stackpos].edgeNode = edgeNode;
-					stack[stackpos].triNode = triNode->kids[1];
-				}
-			} else if (edgeNodeIsectsLeftChild) {
-				// hits only left child
-				++stackpos;
-				stack[stackpos].edgeNode = edgeNode;
-				stack[stackpos].triNode = triNode->kids[0];
-			} else {
-				// hits none
+			// does the edgeNode (with its aabb described in 6 planes transformed and rotated to
+			// b's coordinates) intersect with one or other of b's child nodes?
+			AABBd rotAabb = rotateAaabb(edgeNode->aabb, transTo);
+			const bool left = rotAabb.Intersects(triBvh->GetNode(triNode->kids[0])->aabb);
+			const bool right = rotAabb.Intersects(triBvh->GetNode(triNode->kids[1])->aabb);
+
+			if (left & right) {
+				// Recurse into edge nodes until we find one that's smaller than the tri node
+				// (or hit a single edge leaf)
+				stack[++stackpos] = stackobj { edgeNode->kids[1], curr.triNode };
+				stack[++stackpos] = stackobj { edgeNode->kids[0], curr.triNode };
+			} else if (left) {
+				// Triangle BVH node is larger than the edge, split it and try again
```

Comment:
Please explain why "larger"? It seems that it can be of any size, it just intersected not with two but with one edge node.

Reply:
This is my best interpretation of an existing algorithm that was otherwise undocumented.

This seemed to be the rationale behind the decision to split of intersecting with only one edge node - it tries to ensure both the edge and triangle nodes are of approximately equal size to prevent one large parent BVH node on the left tree triggering an evaluation against all child BVH nodes on the right tree before splitting the left node.